README.md
# Rescalar imágenes API Rest.

Rescalado de imágenes con la librería [Sharp](https://www.npmjs.com/package/sharp/).
Prueba de acceso a Plexus.

---
## Arquitectura

El patrón de arquitectura seguida en el diseño es Modelo–vista–controlador.
En *./src/* encotramos el código de la API.
Dicho proyecto tiene eslint y standard version para un manejo mas rápido del versionado y bug-fixes.

---
## API REST

La API esta compuesta de dos rutas:
- ##### POST /task
    -La ruta espera un FormData con la imagen almacenada en *original*.
    -Devuelve un JSON que contiene el taskId y status,
        -Si la imagen ya existe o no se a recibido devuelve status:"error" y un mensaje (message), sin taskId.
- ##### GET /task/:taksId
    -Devuelve un JSON con:
        -status: "error"|"success",
        -message: Si el estado es erroneo devuelve el motivo (tarea no encontrada).
        -task: Toda la información almacena de la tarea.

## Formato de los JSON

- ##### Tasks.json
    {
        [tasksId]:{
            "800": "complete"|"imcomplete",
            "1024": "complete"|"imcomplete",
            "status": "complete"|"imcomplete",
            "date": timestamp,
            "path": "./images/original/filename",
            "1024Timestamp": timestamp,
            "800Timestamp": timestamp
        }
    .
    .
    .
    }
- ##### Images.json
    {
        [filename]:{
            "date": timestamp,
            "pathOriginal": "./images/original/filename",
            "path800": "./images/output/filename/800/md5.ext",
            "path1024": "./images/output/filename/1024/md5.ext",
            "taskId": tasksId
        }
    .
    .
    .
    }

- #### Instalación y ejecución

      $ npm i
      $ npm run start
*Utilizar [Postman](https://www.postman.com/) para realizar las peticiones.
**Por defecto el servidor se ejecuta en el puerto 3400 se puede cambiar desde *./src/config.js***


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.0.2 (2021-01-15)


### Features

* **images:** save image info in JSON ([a7990da](https://github.com/JavierQ15/resize-image-plexus/commit/a7990da56dd048de074244e24b39e03e2e5e5d5f))
* **task:** get task status ([94b8ac4](https://github.com/JavierQ15/resize-image-plexus/commit/94b8ac40ce47ee464ef61ef885e957a3be70032a))
* **tasks:** storage the status of the task ([8994081](https://github.com/JavierQ15/resize-image-plexus/commit/8994081952f6dc37dfc58555b4ce04b60fc95f84))
* jSON files where persistent information is stored ([6344455](https://github.com/JavierQ15/resize-image-plexus/commit/63444555f7e7984e3c732489ad74504d6b10bd15))
* **task:** generate processed images ([a3a39b8](https://github.com/JavierQ15/resize-image-plexus/commit/a3a39b83e2731a8d60d5df351959d8b5d2518e6b))
* create the main structure of the server ([e07c76a](https://github.com/JavierQ15/resize-image-plexus/commit/e07c76a3b6b89a95ab76d8487d2864404ea8b894))


### Bug Fixes

* **task:** correct path ([613e4fb](https://github.com/JavierQ15/resize-image-plexus/commit/613e4fbd760e01f1fb38844d86397ad8dff57abb))


src/app.js
// Load required libraries
const express = require("express");
const bodyParser = require("body-parser");
const helmet = require("helmet");

// Run express

const app = express();

// Load routes
const taskRoutes = require("./routes/task");

// Middlewares
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json({ limit: "50mb" }));
app.use(helmet());

// CORS
app.use((req, res, next) => {
  res.header("Access-Control-Allow-Origin", "*");
  res.header(
    "Access-Control-Allow-Headers",
    "Authorization, X-API-KEY, Origin, X-Requested-With, Content-Type, Accept, Access-Control-Allow-Request-Method"
  );
  res.header("Access-Control-Allow-Methods", "GET, POST, OPTIONS, PUT, DELETE");
  res.header("Allow", "GET, POST, OPTIONS, PUT, DELETE");
  next();
});

// Add prefixes to routes / load routes
app.use("/task", taskRoutes);

// Export module
module.exports = app;


src/controllers/task.js
//Image processing library
const sharp = require("sharp");
//Libraries to work with files
const fs = require("fs");
const path = require("path");
//Library to generate md5
const md5 = require("crypto-js/md5");
//JSON to storage the information
const TasksPath = path.resolve("./src/Tasks.json");
let Tasks = JSON.parse(fs.readFileSync(TasksPath));
const ImagesPath = path.resolve("./src/Images.json");
let Images = JSON.parse(fs.readFileSync(ImagesPath));
module.exports = {
    processImage: (req, res) => {
        const file = req.file;
        if (!file)
            return res
                .status(400)
                .send({ status: "error", message: "We have not received any images" });
        if(Images[file.originalname]) return res.status(409).send({status:"error",message:"An image with the same name has already been processed"})
        //Generate random taskId
        const taskId =
            Math.random().toString(36).substr(2, 9) +
            Math.random().toString(36).substr(2, 9);
        //File name without extension
        const filename = file.originalname.split(".").slice(0, -1).join(".");
        let newTasks = {
            status: "incomplete",
            date: Date.now(),
            1024: "imcomplete",
            800: "imcomplete",
            path: `./images/original/${filename}.${file.originalname.replace(/^.*[.]/, "")}`,
        };
        let newImages = {
            date: Date.now(),
            pathOriginal: `./images/original/${filename}.${file.originalname.replace(/^.*[.]/, "")}`,
            path800: `./images/output/${filename}/800/${md5(
                filename + "_800"
            )}.${file.originalname.replace(/^.*[.]/, "")}`,
            path1024:`./images/output/${filename}/1024/${md5(
                filename + "_1024"
            )}.${file.originalname.replace(/^.*[.]/, "")}`,
            taskId
        }
        Tasks = { ...Tasks, [taskId]: newTasks };
        fs.writeFileSync(TasksPath, JSON.stringify(Tasks, null, 4));
        Images = { ...Images, [file.originalname]: newImages };
        fs.writeFileSync(ImagesPath, JSON.stringify(Images, null, 4));
        const imageBuffer = fs.readFileSync(file.path);
        sharp(imageBuffer)
            .resize({ width: 1024 })
            .toFile(
                path.resolve(
                    newImages.path1024
                )
            )
            .then(() => {
                newTasks["1024"] = "complete";
                newTasks["1024Timestamp"] = Date.now();
                Tasks = { ...Tasks, [taskId]: newTasks };
                fs.writeFileSync(TasksPath, JSON.stringify(Tasks, null, 4));
                sharp(imageBuffer)
                    .resize({ width: 800 })
                    .toFile(
                        path.resolve(
                            newImages.path800
                        )
                    )
                    .then(() => {
                        newTasks["800"] = "complete";
                        newTasks["800Timestamp"] = Date.now();
                        newTasks["status"] = "complete";
                        Tasks = { ...Tasks, [taskId]: newTasks };
                        fs.writeFileSync(TasksPath, JSON.stringify(Tasks, null, 4));
                    });
            });
        res.send({ status: "success", taskId });
    },
    getStatus: (req, res) => {
        const {taskId} = req.params
        const task = Tasks[taskId]
        if (!task)
            return res
                .status(400)
                .send({ status: "error", message: "We have not found any task with that ID" });
        return res.send({status: "success",task});
    }
};
